src/storage/src/render/upsert.rs

```diff
@@ -211,22 +217,47 @@ where
                         a_ts == b_ts && a_key == b_key
                     });
 
+                    // Read the previous values _per key_ out of `state`, recording it
+                    // along with the value with the _latest timestamp for that key_.
+                    commands_state.clear();
                     while let Some((ts, key, _, value)) = commands.next() {
-                        match value {
+                        let previous_value = state.get(&key).cloned();
+                        commands_state.insert(key, (value, previous_value.clone(), ts));
```

Comment:
@moulimukherjee this pr also makes it so we dont need to think about ordering in the upsert trait impls, we can just do it here!

Reply:
This is great \o/!

Reply:
This will throw away commands for the same key but different times but we need to produce these updates.

Reply:
true, will update